### ISSUE TITLE
Support for asynchronous User-Defined Functions (Excel 2010)

### ISSUE THREAD

**Issue by Juanlu001** [ZoomerAnalytics/xlwings#473]
Excel 2010 introduced asynchronous User-Defined Functions, which can be extremely useful for long-running tasks:

https://msdn.microsoft.com/en-us/library/office/ff796219%28v=office.14%29.aspx

I noticed that [@ericremoreynolds commented async UDFs were on your roadmap](https://github.com/ZoomerAnalytics/xlwings/issues/362#issuecomment-188130042), but I still wanted to open an issue to track this specific request and perhaps know if you have an ETA in mind.

I guess [the new async constructs in Python 3.5](https://docs.python.org/3/whatsnew/3.5.html#whatsnew-pep-492) would be a nice starting point, but maybe others can come up with an implementation which is compatible with [Legacy Python](http://carreau.github.io/posts/planning-an-early-death-for-python-2.html) as well.

cc @jmatiasindizen @skasi7

**Comment by denfromufa** [comment]
+1

**Comment by akapocsi** [comment]
+1 for async as well

**Comment by gnillev** [comment]
+1

**Comment by adrianschneider94** [comment]
+1

**Comment by fzumstein** [comment]
Starting with 0.13.0 you can use something like below. It's not totally beautiful yet as the caller is really a native pywin32 object and not an xlwings one. Note that before 0.13.0, threading would cause an issue. Also, this approach will cause volatile functions to be recalculated upon the return of the async function (this could be improved via [RTD](https://github.com/mhammond/pywin32/blob/d64fac8d7bda2cb1d81e2c9366daf99e802e327f/com/win32com/demos/excelRTDServer.py)) but for most use cases the below might be good enough.

```
import xlwings as xw
from threading import Thread
import time


cache = {}

class HelloThread(Thread):
    def __init__(self, book, sheet, address, i):
        Thread.__init__(self)
        self.book = book
        self.sheet = sheet
        self.address = address
        self.i = i

    def run(self):
        # The actual time consuming calculation
        time.sleep(3)
        cache['hello' + str(self.i)] = 2 * self.i

        xw.Book(self.book).sheets[self.sheet].range(self.address).formula = \
            xw.Book(self.book).sheets[self.sheet].range(self.address).formula


@xw.func
@xw.arg('xl', vba='Application')
def hello(i, xl):
    cached_value = cache.get('hello' + str(i))
    if cached_value:
        del cache['hello' + str(i)]
        return cached_value
    else:
        # You can't pass pywin32 objects directly to threads
        thread = HelloThread(xl.Caller.Worksheet.Parent.Name,
                             xl.Caller.Worksheet.Name,
                             xl.Caller.Address,
                             i)
        thread.start()
    return "#N/A"
```

**Comment by fzumstein** [comment]
@Colin-b weren't you using something like this, too? What was your approach?

**Comment by Colin-b** [comment]
Your idea seems better than mine, triggering a new call to the function once result is available :)

On my end I am sending back results once available via: xlwings.server.add_idle_task(DelayWrite(excel_range, {'expand': 'table'}, result, (nb_rows, nb_columns)))

**Comment by fzumstein** [comment]
ok but you were also using a threading/caching mechanism?

**Comment by Colin-b** [comment]
Yes I am starting a thread, this tread will send result when available. And right after starting thread I am sending a temporary result (same as your '#N/A').
I have a caching mechanism as well to avoid useless re-computation, I use cachetools to do it though as I need a TTL parameter to get rid of old results automatically.

**Comment by fzumstein** [comment]
released with v0.14.0
